File: tests/native/org.eclipse.ice.reactor.sfr.native.test/SFRDataTester.cpp
/*******************************************************************************
 * Copyright (c) 2013, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE Regression
#include <boost/test/included/unit_test.hpp>
#include <SFRData.h>
#include <string>
#include <vector>
#include <float.h>
#include <memory>

using namespace ICE_SFReactor;

BOOST_AUTO_TEST_SUITE (SFRDataTester_testSuite)

BOOST_AUTO_TEST_CASE(checkConstruction) {
	// begin-user-code

	// Initialize a new SFRData
	SFRData data;

	// Default values:
	std::string feature = "Feature 1";

	std::vector<double> position = { 0.0, 0.0, 0.0 };

	double uncertainty = 0.0;
	std::string units = "seconds";
	double value = 0.0;

	/* ---- Check nullary constructor. ---- */

	BOOST_REQUIRE_EQUAL(feature, data.getFeature());
	BOOST_REQUIRE(position == data.getPosition());
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());
	BOOST_REQUIRE_EQUAL(units, data.getUnits());
	BOOST_REQUIRE_EQUAL(value, data.getValue());

	/* ------------------------------------ */

	/* ---- Check feature-based constructor. ---- */

	data = SFRData("Infinite Improbability Device");

	BOOST_REQUIRE_EQUAL("Infinite Improbability Device", data.getFeature());
	BOOST_REQUIRE(position == data.getPosition());
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());
	BOOST_REQUIRE_EQUAL(units, data.getUnits());
	BOOST_REQUIRE_EQUAL(value, data.getValue());
	/* ------------------------------------------ */

	/* ---- Check invalid constructor (empty string). ---- */
	data = SFRData("");

	BOOST_REQUIRE_EQUAL(feature, data.getFeature());
	BOOST_REQUIRE(position == data.getPosition());
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());
	BOOST_REQUIRE_EQUAL(units, data.getUnits());
	BOOST_REQUIRE_EQUAL(value, data.getValue());

	/* ------------------------------------------- */

	/* ---- Check invalid constructor (blank). ---- */
	data = SFRData("   ");

	BOOST_REQUIRE_EQUAL(feature, data.getFeature());
	BOOST_REQUIRE(position == data.getPosition());
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());
	BOOST_REQUIRE_EQUAL(units, data.getUnits());
	BOOST_REQUIRE_EQUAL(value, data.getValue());
	/* -------------------------------------------- */

	return;
	// end-user-code
}

BOOST_AUTO_TEST_CASE(checkValue) {
	// begin-user-code

	// Initialize a new SFRData.
	SFRData data;

	// Check the default value.
	double value = 0.0;
	BOOST_REQUIRE_EQUAL(value, data.getValue());
	data.setValue(0.0);
	BOOST_REQUIRE_EQUAL(value, data.getValue());

	// Update the value.
	value = 42.0;
	data.setValue(42.0);
	BOOST_REQUIRE_EQUAL(value, data.getValue());

	// Try setting it to some big values.
	value = DBL_MAX;
	data.setValue(DBL_MAX);
	BOOST_REQUIRE_EQUAL(value, data.getValue());

	// Try setting it to some really small values.
	value = DBL_MIN;
	data.setValue(DBL_MIN);
	BOOST_REQUIRE_EQUAL(value, data.getValue());

	return;
	// end-user-code
}

BOOST_AUTO_TEST_CASE(checkUncertainty) {
	// begin-user-code

	// Initialize a new SFRData.
	SFRData data;

	// Check the default uncertainty.
	double uncertainty = 0.0;
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());
	data.setUncertainty(0.0);
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());

	// Update the value.
	uncertainty = 42.0;
	data.setUncertainty(42.0);
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());

	// Try setting it to some big values.
	uncertainty = DBL_MAX;
	data.setUncertainty(DBL_MAX);
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());

	// Try setting it to some really small values.
	uncertainty = DBL_MIN;
	data.setUncertainty(DBL_MIN);
	BOOST_REQUIRE_EQUAL(uncertainty, data.getUncertainty());

	return;
// end-user-code
}

BOOST_AUTO_TEST_CASE(checkUnits) {
	// begin-user-code

	// Initialize a new SFRData.
	SFRData data;

	std::string defaultstring = "seconds";
	std::string normalstring = "minutes";
	std::string emptystring = "";
	std::string whitespacestring = "  	";
	std::string normalstringPlus = " minutes 	";

	// Check the default.
	BOOST_REQUIRE_EQUAL(defaultstring, data.getUnits());

	// Check setting it to an empty string.
	data.setUnits(emptystring);
	BOOST_REQUIRE_EQUAL(defaultstring, data.getUnits());

	// Check setting it to a string with whitespace only.
	data.setUnits(whitespacestring);
	BOOST_REQUIRE_EQUAL(defaultstring, data.getUnits());

	// Check setting it to a normal string.
	data.setUnits(normalstring);
	BOOST_REQUIRE_EQUAL(normalstring, data.getUnits());

	// Revert it to default temporarily.
	data.setUnits(defaultstring);
	BOOST_REQUIRE_EQUAL(defaultstring, data.getUnits());

	// Check setting it to a normal string with whitespace.
	data.setUnits(normalstringPlus);
	BOOST_REQUIRE_EQUAL(normalstring, data.getUnits());

	return;
	// end-user-code
}

BOOST_AUTO_TEST_CASE(checkFeature) {
	// begin-user-code

	// Initialize a new SFRdata.
	SFRData data;

	std::string defaultstring = "Feature 1";
	std::string normalstring = "Dimensions Traversed";
	std::string emptystring = "";
	std::string whitespacestring = "  	";
	std::string normalstringPlus = " Dimensions Traversed 	";

	// Check the default.
	BOOST_REQUIRE_EQUAL(defaultstring, data.getFeature());

	// Check setting it to an empty string.
	data.setFeature(emptystring);
	BOOST_REQUIRE_EQUAL(defaultstring, data.getFeature());

	// Check setting it to a string with whitespace only.
	data.setFeature(whitespacestring);
	BOOST_REQUIRE_EQUAL(defaultstring, data.getFeature());

	// Check setting it to a normal string.
	data.setFeature(normalstring);
	BOOST_REQUIRE_EQUAL(normalstring, data.getFeature());

	// Revert it to default temporarily.
	data.setFeature(defaultstring);
	BOOST_REQUIRE_EQUAL(defaultstring, data.getFeature());

	// Check setting it to a normal string with whitespace.
	data.setFeature(normalstringPlus);
	BOOST_REQUIRE_EQUAL(normalstring, data.getFeature());

	return;
// end-user-code
}
BOOST_AUTO_TEST_CASE(checkPosition) {

	// begin-user-code

	// Initialize a new SFRdata
	std::unique_ptr<SFRData> data (new SFRData());

	// Initialize the default position.
	std::vector<double> defaultPosition;
	defaultPosition.push_back(0.0);
	defaultPosition.push_back(0.0);
	defaultPosition.push_back(0.0);

	// Make sure the default position is set.
	BOOST_REQUIRE_EQUAL(3, data.get()->getPosition().size());
	BOOST_REQUIRE_EQUAL(defaultPosition.at(0), data.get()->getPosition().at(0));
	BOOST_REQUIRE_EQUAL(defaultPosition.at(1), data.get()->getPosition().at(1));
	BOOST_REQUIRE_EQUAL(defaultPosition.at(2), data.get()->getPosition().at(2));

	// Initialize a non-default position.
	std::vector<double> position;

	// Add a position value. The list size is 1 and should not change the
	// data's position.
	position.push_back(1.0);
	data.get()->setPosition(position);
	BOOST_REQUIRE_EQUAL(3, data.get()->getPosition().size());
	BOOST_REQUIRE_EQUAL(defaultPosition.at(0), data.get()->getPosition().at(0));
	BOOST_REQUIRE_EQUAL(defaultPosition.at(1), data.get()->getPosition().at(1));
	BOOST_REQUIRE_EQUAL(defaultPosition.at(2), data.get()->getPosition().at(2));

	// Add a position value. The list size is 2 and should not change the
	// data's position.
	position.push_back(-1.0);
	data.get()->setPosition(position);
	BOOST_REQUIRE_EQUAL(3, data.get()->getPosition().size());
	BOOST_REQUIRE_EQUAL(defaultPosition.at(0), data.get()->getPosition().at(0));
	BOOST_REQUIRE_EQUAL(defaultPosition.at(1), data.get()->getPosition().at(1));
	BOOST_REQUIRE_EQUAL(defaultPosition.at(2), data.get()->getPosition().at(2));

	// Add a position value. The list size is 3 and should change the data's
	// position.
	position.push_back(-1.0);
	data.get()->setPosition(position);
	BOOST_REQUIRE_EQUAL(3, data.get()->getPosition().size());
	BOOST_REQUIRE_EQUAL(position.at(0), data.get()->getPosition().at(0));
	BOOST_REQUIRE_EQUAL(position.at(1), data.get()->getPosition().at(1));
	BOOST_REQUIRE_EQUAL(position.at(2), data.get()->getPosition().at(2));

	// Add a fourth position value. The list size is 4 and should not change
	// the data's position.
	position.push_back(-1.0);
	data.get()->setPosition(position);
	BOOST_REQUIRE_EQUAL(3, data.get()->getPosition().size());
	BOOST_REQUIRE_EQUAL(position.at(0), data.get()->getPosition().at(0));
	BOOST_REQUIRE_EQUAL(position.at(1), data.get()->getPosition().at(1));
	BOOST_REQUIRE_EQUAL(position.at(2), data.get()->getPosition().at(2));

	// Remove the fourth element from our position.
	position.pop_back();
	BOOST_REQUIRE_EQUAL(3, data.get()->getPosition().size());
	BOOST_REQUIRE_EQUAL(position.at(0), data.get()->getPosition().at(0));
	BOOST_REQUIRE_EQUAL(position.at(1), data.get()->getPosition().at(1));
	BOOST_REQUIRE_EQUAL(position.at(2), data.get()->getPosition().at(2));

	// Set it to an empty vector. This should not change the position.
	std::vector<double> emptyVector;
	data.get()->setPosition(emptyVector);
	BOOST_REQUIRE_EQUAL(3, data.get()->getPosition().size());
	BOOST_REQUIRE_EQUAL(position.at(0), data.get()->getPosition().at(0));
	BOOST_REQUIRE_EQUAL(position.at(1), data.get()->getPosition().at(1));
	BOOST_REQUIRE_EQUAL(position.at(2), data.get()->getPosition().at(2));

	return;
// end-user-code
}

BOOST_AUTO_TEST_CASE(checkEquality) {
	// begin-user-code

	SFRData object, equalObject, unequalObject, invalidObject;

	// Some values for loading into the data.
	std::string feature = "intelligence";
	std::string nullFeature;
	std::string units = "mouse brain";
	double value = 1.0;
	double uncertainty = 0.0;
	std::vector<double> position;
	position.push_back(1.0);
	position.push_back(2.0);
	position.push_back(3.0);

	// Set the information for the initial data.
	object = SFRData("Mouse");
	object.setFeature(feature);
	object.setUnits(units);
	object.setValue(value);
	object.setUncertainty(uncertainty);
	object.setPosition(position);

	// Set the information for the equal data.
	equalObject = SFRData("Mouse");
	equalObject.setFeature(feature);
	equalObject.setUnits(units);
	equalObject.setValue(value);
	equalObject.setUncertainty(uncertainty);
	equalObject.setPosition(position);

	// Set the information for the unequal data.
	unequalObject = SFRData("Human");
	unequalObject.setFeature(feature);
	unequalObject.setUnits(units);
	unequalObject.setValue(0.5);
	unequalObject.setUncertainty(0.25);
	unequalObject.setPosition(position);

	// Test reflexivity and symmetry.
	BOOST_REQUIRE(object == object);
	BOOST_REQUIRE(object == equalObject);
	BOOST_REQUIRE(equalObject == object);

	// Test inequality.
	BOOST_REQUIRE(!(object == unequalObject));
	BOOST_REQUIRE(!(unequalObject == object));

	// Test inequality with invalid objects.
	BOOST_REQUIRE(!(object == invalidObject));
	BOOST_REQUIRE(!(invalidObject == object));

	return;
	// end-user-code
}

BOOST_AUTO_TEST_CASE(checkCopying) {
	// begin-user-code

	// Some defaults to use.
	std::string feature = "Acceleration";
	std::string units = "ft/s/s";
	double value = 32.0;

	// Create a new SFRData with some special features and data
	SFRData object("Bowl of petunias");
	object.setFeature(feature);
	object.setUnits(units);
	object.setValue(value);

	std::vector<double> position;
	position.push_back(42.0);
	position.push_back(42000.0);
	position.push_back(84.0);
	object.setPosition(position);

	// Create objects for copying/cloning.
	SFRData copiedObject;
	std::shared_ptr<IData> clonedObject;

	/* ---- Test copying. ---- */
	// Make sure these are different objects beforehand!
	BOOST_REQUIRE(!(object == copiedObject));

	// Copy the contents of object into copiedObject.
	copiedObject = SFRData(object);

	// Check that they are the same now
	BOOST_REQUIRE(object == copiedObject);

	// Make sure we didn't clobber the old contents of object.
	BOOST_REQUIRE_EQUAL(object.getFeature(), feature);
	BOOST_REQUIRE_EQUAL(object.getUnits(), units);
	BOOST_REQUIRE_EQUAL(object.getValue(), value);
	BOOST_REQUIRE(object.getPosition() == position);
	/* ----------------------- */

	/* ---- Test cloning. ---- */

	// Clone the object.
	clonedObject = object.clone();

	// Check that the clone isn't null
	BOOST_REQUIRE(clonedObject);

	// Check that they are the same
	BOOST_REQUIRE(object == *clonedObject);

	// Make sure we didn't clobber the old contents of object.
	BOOST_REQUIRE_EQUAL(object.getFeature(), feature);
	BOOST_REQUIRE_EQUAL(object.getUnits(), units);
	BOOST_REQUIRE_EQUAL(object.getValue(), value);
	BOOST_REQUIRE(object.getPosition() == position);
	/* ----------------------- */

	return;
// end-user-code
}
BOOST_AUTO_TEST_SUITE_END()

File: tests/org.eclipse.ice.reactor.sfr.test/src/org/eclipse/ice/reactor/sfr/test/SFRDataTester.java
/*******************************************************************************
 * Copyright (c) 2013, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
package org.eclipse.ice.reactor.sfr.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.util.ArrayList;
import java.util.List;

import org.eclipse.ice.reactor.sfr.base.SFRData;
import org.junit.Test;

/**
 * <p>
 * Tests the operations of the SFRData class.
 * </p>
 * 
 * @author Anna Wojtowicz
 */
public class SFRDataTester {

	/**
	 * <p>
	 * Tests the constructors and default values of the SFRData class.
	 * </p>
	 * 
	 */
	@Test
	public void checkConstruction() {

		SFRData data;

		// Default values:
		String feature = "Feature 1";

		List<Double> position = new ArrayList<Double>(3);
		position.add(0.0);
		position.add(0.0);
		position.add(0.0);

		double uncertainty = 0.0;
		String units = "seconds";
		double value = 0.0;

		/* ---- Check nullary constructor. ---- */
		data = new SFRData();

		assertEquals(feature, data.getFeature());
		assertEquals(position, data.getPosition());
		assertEquals(uncertainty, data.getUncertainty(), 0);
		assertEquals(units, data.getUnits());
		assertEquals(value, data.getValue(), 0);
		/* ------------------------------------ */

		/* ---- Check feature-based constructor. ---- */
		data = new SFRData("Infinite Improbability Device");

		assertEquals("Infinite Improbability Device", data.getFeature());
		assertEquals(position, data.getPosition());
		assertEquals(uncertainty, data.getUncertainty(), 0);
		assertEquals(units, data.getUnits());
		assertEquals(value, data.getValue(), 0);
		/* ------------------------------------------ */

		/* ---- Check invalid constructor (null). ---- */
		data = new SFRData(null);

		assertEquals(feature, data.getFeature());
		assertEquals(position, data.getPosition());
		assertEquals(uncertainty, data.getUncertainty(), 0);
		assertEquals(units, data.getUnits());
		assertEquals(value, data.getValue(), 0);
		/* ------------------------------------------- */

		/* ---- Check invalid constructor (blank). ---- */
		data = new SFRData("   ");

		assertEquals(feature, data.getFeature());
		assertEquals(position, data.getPosition());
		assertEquals(uncertainty, data.getUncertainty(), 0);
		assertEquals(units, data.getUnits());
		assertEquals(value, data.getValue(), 0);
		/* -------------------------------------------- */

		return;
	}

	/**
	 * <p>
	 * Tests the getter and setter of the value attribute.
	 * </p>
	 * 
	 */
	@Test
	public void checkValue() {

		// Initialize a new SFRData.
		SFRData data = new SFRData();

		// Check the default value.
		double value = 0.0;
		assertEquals(value, data.getValue(), 0);
		data.setValue(0.0);
		assertEquals(value, data.getValue(), 0);

		// Update the value.
		value = 42.0;
		data.setValue(42.0);
		assertEquals(value, data.getValue(), 0);

		// Try setting it to some big values.
		value = Double.MAX_VALUE;
		data.setValue(Double.MAX_VALUE);
		assertEquals(value, data.getValue(), 0);

		// Try setting it to some really small values.
		value = Double.MIN_VALUE;
		data.setValue(Double.MIN_VALUE);
		assertEquals(value, data.getValue(), 0);

		return;
	}

	/**
	 * <p>
	 * Tests the getter and setter for the uncertainty attribute.
	 * </p>
	 * 
	 */
	@Test
	public void checkUncertainty() {

		// Initialize a new SFRData.
		SFRData data = new SFRData();

		// Check the default uncertainty.
		double uncertainty = 0.0;
		assertEquals(uncertainty, data.getUncertainty(), 0);
		data.setUncertainty(0.0);
		assertEquals(uncertainty, data.getUncertainty(), 0);

		// Update the value.
		uncertainty = 42.0;
		data.setUncertainty(42.0);
		assertEquals(uncertainty, data.getUncertainty(), 0);

		// Try setting it to some big values.
		uncertainty = Double.MAX_VALUE;
		data.setUncertainty(Double.MAX_VALUE);
		assertEquals(uncertainty, data.getUncertainty(), 0);

		// Try setting it to some really small values.
		uncertainty = Double.MIN_VALUE;
		data.setUncertainty(Double.MIN_VALUE);
		assertEquals(uncertainty, data.getUncertainty(), 0);

		return;
	}

	/**
	 * <p>
	 * Tests the getter and setter of the units attribute.
	 * </p>
	 * 
	 */
	@Test
	public void checkUnits() {

		// Initialize a new SFRData.
		SFRData data = new SFRData();

		String defaultString = "seconds";
		String normalString = "minutes";
		String emptyString = "";
		String whitespaceString = "  	";
		String normalStringPlus = " minutes 	";

		// Check the default.
		assertEquals(defaultString, data.getUnits());

		// Check setting it to a null String.
		data.setUnits(null);
		assertEquals(defaultString, data.getUnits());

		// Check setting it to an empty String.
		data.setUnits(emptyString);
		assertEquals(defaultString, data.getUnits());

		// Check setting it to a String with whitespace only.
		data.setUnits(whitespaceString);
		assertEquals(defaultString, data.getUnits());

		// Check setting it to a normal String.
		data.setUnits(normalString);
		assertEquals(normalString, data.getUnits());

		// Revert it to default temporarily.
		data.setUnits(defaultString);
		assertEquals(defaultString, data.getUnits());

		// Check setting it to a normal String with whitespace.
		data.setUnits(normalStringPlus);
		assertEquals(normalString, data.getUnits());

		return;
	}

	/**
	 * <p>
	 * Tests the getter and setter of the feature attribute.
	 * </p>
	 * 
	 */
	@Test
	public void checkFeature() {

		// Initialize a new SFRData.
		SFRData data = new SFRData();

		String defaultString = "Feature 1";
		String normalString = "Dimensions Traversed";
		String emptyString = "";
		String whitespaceString = "  	";
		String normalStringPlus = " Dimensions Traversed 	";

		// Check the default.
		assertEquals(defaultString, data.getFeature());

		// Check setting it to a null String.
		data.setFeature(null);
		assertEquals(defaultString, data.getFeature());

		// Check setting it to an empty String.
		data.setFeature(emptyString);
		assertEquals(defaultString, data.getFeature());

		// Check setting it to a String with whitespace only.
		data.setFeature(whitespaceString);
		assertEquals(defaultString, data.getFeature());

		// Check setting it to a normal String.
		data.setFeature(normalString);
		assertEquals(normalString, data.getFeature());

		// Revert it to default temporarily.
		data.setFeature(defaultString);
		assertEquals(defaultString, data.getFeature());

		// Check setting it to a normal String with whitespace.
		data.setFeature(normalStringPlus);
		assertEquals(normalString, data.getFeature());

		return;
	}

	/**
	 * <p>
	 * Tests the getter and setter for the position attribute.
	 * </p>
	 * 
	 */
	@Test
	public void checkPosition() {

		// Initialize a new SFRData.
		SFRData data = new SFRData();

		// Initialize the default position.
		ArrayList<Double> defaultPosition = new ArrayList<Double>();
		defaultPosition.add(0.0);
		defaultPosition.add(0.0);
		defaultPosition.add(0.0);

		// Make sure the default position is set.
		assertEquals(defaultPosition, data.getPosition());

		// Initialize a non-default position.
		ArrayList<Double> position = new ArrayList<Double>();

		// Add a position value. The list size is 1 and should not change the
		// data's position.
		position.add(1.0);
		data.setPosition(position);
		assertEquals(defaultPosition, data.getPosition());

		// Add a position value. The list size is 2 and should not change the
		// data's position.
		position.add(-1.0);
		data.setPosition(position);
		assertEquals(defaultPosition, data.getPosition());

		// Add a position value. The list size is 3 and should change the data's
		// position.
		position.add(-1.0);
		data.setPosition(position);
		assertEquals(position, data.getPosition());

		// Add a fourth position value. The list size is 4 and should not change
		// the data's position.
		position.add(-1.0);
		data.setPosition(position);
		assertFalse(position.equals(data.getPosition()));
		assertEquals(position.subList(0, 3), data.getPosition());

		// Remove the fourth element from our position.
		position.remove(3);
		assertEquals(position, data.getPosition());

		// Set it to an empty ArrayList. This should not change the position.
		data.setPosition(new ArrayList<Double>());
		assertEquals(position, data.getPosition());

		// Set it to null. This should not change the position.
		data.setPosition(null);
		assertEquals(position, data.getPosition());

		return;
	}

	/**
	 * <p>
	 * Tests the equality operation of SFRData.
	 * </p>
	 */
	@Test
	public void checkEquality() {

		SFRData object, equalObject, unequalObject, invalidObject;

		// Some values for loading into the data.
		String units = "mouse brain";
		double value = 1.0;
		double uncertainty = 0.0;
		ArrayList<Double> position = new ArrayList<Double>();
		position.add(1.0);
		position.add(2.0);
		position.add(3.0);

		// Set the information for the initial data.
		object = new SFRData("Mouse");
		object.setUnits(units);
		object.setValue(value);
		object.setUncertainty(uncertainty);
		object.setPosition(position);

		// Set the information for the equal data.
		equalObject = new SFRData("Mouse");
		equalObject.setUnits(units);
		equalObject.setValue(value);
		equalObject.setUncertainty(uncertainty);
		equalObject.setPosition(position);

		// Set the information for the unequal data.
		unequalObject = new SFRData("Human");
		unequalObject.setUnits(units);
		unequalObject.setValue(0.5);
		unequalObject.setUncertainty(0.25);
		unequalObject.setPosition(position);

		// Set the information for an invalid SFRData.
		invalidObject = new SFRData(null);
		invalidObject.setUnits(units);
		invalidObject.setValue(value);
		invalidObject.setUncertainty(uncertainty);
		invalidObject.setPosition(position);

		// Test reflexivity and symmetry.
		assertTrue(object.equals(object));
		assertTrue(object.equals(equalObject));
		assertTrue(equalObject.equals(object));

		// Test hash code equality.
		assertTrue(object.hashCode() == object.hashCode());
		assertTrue(object.hashCode() == equalObject.hashCode());

		// Test inequality.
		assertFalse(object.equals(unequalObject));
		assertFalse(unequalObject.equals(object));

		// Test hash code inequality.
		assertFalse(object.hashCode() == unequalObject.hashCode());

		// Test inequality with invalid objects.
		assertFalse(object.equals(invalidObject));
		assertFalse(invalidObject.equals(object));

		// Test inequality with null values
		assertFalse(object==null);
		assertFalse(equalObject==null);
		assertFalse(unequalObject==null);
		assertFalse(invalidObject==null);

		// Test inequality with bad values.
		assertFalse(object.equals(42));
		assertFalse("just a string".equals(object));

		return;
	}

	/**
	 * <p>
	 * Tests the copying and cloning methods of SFRData.
	 * </p>
	 */
	@Test
	public void checkCopying() {

		String feature = "Acceleration";
		String units = "ft/s/s";
		double value = 32.0;

		// Create a new SFRData with some special features and data.
		SFRData object = new SFRData("Bowl of petunias");
		object.setFeature(feature);
		object.setUnits(units);
		object.setValue(value);

		ArrayList<Double> position = new ArrayList<Double>();
		position.add(42.0);
		position.add(42000.0);
		position.add(84.0);
		object.setPosition(position);

		// Create objects for copying/cloning.
		SFRData copiedObject = new SFRData("Sperm whale");
		SFRData clonedObject = null;

		/* ---- Test passing in bad arguments to copy. ---- */
		object.copy(null);

		// Make sure we didn't clobber the old contents of object.
		assertTrue(object.getFeature().equals(feature));
		assertTrue(object.getUnits().equals(units));
		assertTrue(object.getValue() == value);
		assertTrue(object.getPosition().equals(position));
		/* ------------------------------------------------ */

		/* ---- Test copying. ---- */
		// Make sure these are different objects beforehand!
		assertFalse(object == copiedObject);
		assertFalse(object.equals(copiedObject));

		// Copy the contents of object into copiedObject.
		copiedObject.copy(object);

		// Make sure they are still different references but have the same info.
		assertFalse(object == copiedObject);
		assertTrue(object.equals(copiedObject));

		// Make sure we didn't clobber the old contents of object.
		assertTrue(object.getFeature().equals(feature));
		assertTrue(object.getUnits().equals(units));
		assertTrue(object.getValue() == value);
		assertTrue(object.getPosition().equals(position));
		/* ----------------------- */

		/* ---- Test cloning. ---- */
		// Make sure these are different objects beforehand!
		assertFalse(object.equals(clonedObject));

		// Clone the object.
		clonedObject = (SFRData) object.clone();

		// Make sure they are still different references but have the same info.
		assertFalse(object == clonedObject);
		assertTrue(object.equals(clonedObject));
		assertFalse(copiedObject == clonedObject);
		assertTrue(copiedObject.equals(clonedObject));

		// Make sure we didn't clobber the old contents of object.
		assertTrue(object.getFeature().equals(feature));
		assertTrue(object.getUnits().equals(units));
		assertTrue(object.getValue() == value);
		assertTrue(object.getPosition().equals(position));
		/* ----------------------- */

		return;
	}
}